pkg/distributor/distributor_test.go

```diff
@@ -4757,6 +4759,64 @@ func TestSeriesAreShardedToCorrectIngesters(t *testing.T) {
 	assert.Equal(t, series, totalMetadata) // each series has unique metric name, and each metric name gets metadata
 }
 
+func TestHandleIngesterPushError(t *testing.T) {
+	testErrorMsg := "this is a test error message"
+	outputErrorMsgPrefix := "failed pushing to ingester"
+	userID := "test"
+	errWithUserID := fmt.Errorf("user=%s: %s", userID, testErrorMsg)
+	test := map[string]struct {
+		ingesterPushError   error
+		expectedOutputError error
+	}{
+		"no error gives no error": {
+			ingesterPushError:   nil,
+			expectedOutputError: nil,
+		},
+		"an http 400 error gives an http 400 error": {
+			ingesterPushError:   httpgrpc.Errorf(http.StatusBadRequest, testErrorMsg),
+			expectedOutputError: httpgrpc.Errorf(http.StatusBadRequest, "%s: %s", outputErrorMsgPrefix, testErrorMsg),
+		},
+		"an http 500 error gives an http 500 error": {
+			ingesterPushError:   httpgrpc.Errorf(http.StatusInternalServerError, testErrorMsg),
+			expectedOutputError: httpgrpc.Errorf(http.StatusInternalServerError, "%s: %s", outputErrorMsgPrefix, testErrorMsg),
+		},
+		"grpc errMaxInflightRequestsReached error gives a wrapped errMaxInflightRequestsReached error": {
+			ingesterPushError:   errMaxInflightRequestsReached,
+			expectedOutputError: errors.Wrap(errMaxInflightRequestsReached, outputErrorMsgPrefix),
+		},
+		"grpc errMaxIngestionRateReached error gives a wrapped errMaxIngestionRateReached error": {
+			ingesterPushError:   errMaxIngestionRateReached,
+			expectedOutputError: errors.Wrap(errMaxIngestionRateReached, outputErrorMsgPrefix),
+		},
+		"a random ingester error without status gives the same wrapped errpr": {
+			ingesterPushError:   fmt.Errorf(testErrorMsg),
+			expectedOutputError: errors.Wrap(fmt.Errorf(testErrorMsg), outputErrorMsgPrefix),
+		},
+		"an ingester error with status 5xx gives an http 5xx error": {
+			ingesterPushError:   ingester.NewErrorWithStatus(errWithUserID, http.StatusServiceUnavailable),
```

Comment:
Distributor will never receive this type. We can test with `status.Error()` instead.